### ISSUE TITLE
Allow to specify ephemeris download folder

### ISSUE THREAD

**Issue by neothemachine** [skyfielders/python-skyfield#73]
With the new way of loading ephemeris data, it loads the files to the "current folder", which may work in most cases but definitely not in all. I think it would be good to have another optional argument for load() where you can specify the path to a download folder. This is especially important for more production-like environments where you cannot control what the "current directory" is from which a script is run, and then the permissions might be set to prevent any creation of files in that one.

**Comment by brandon-rhodes** [comment]
Agreed, that’s a great idea! I'll include something in the very next version.

**Comment by brandon-rhodes** [comment]
Do you have an opinion about the syntax? What would you consider the pros and cons of:
```
load('de430.bsp', directory='~/cache')
```
versus just being able to include it in the filename:
```
load('~/cache/de430.bsp')
```
?

**Comment by neothemachine** [comment]
I would opt for the first version, I find it less confusing. The second 
one somehow merges the actual file location with the remote download 
url/filename. I am tempted to say "hmm, ok, I actually want it to be 
'~/cache/de430.ephemeris'" and then you start looking for the actual 
conceptual ephemeris identifier in that load() function and realize it 
doesn't exist. By separating the two you make it clear that the folder 
(could rename the argument even to "cacheDirectory='..'") is the 
physical location on your disk, and the first argument the ephemeris 
identifier, which happens to be/become the filename. Apart from that, 
cache folder paths are typically stored in some configuration file (for 
more production-like settings) and then you can directly use that 
without having to think about concatenation and slashes etc.

**Comment by tritium21** [comment]
The second jives with most of the other APIs in python that accept paths.  unqualified, paths are assumed to be relative to cwd.  ...The second one actually can be passed unmolested to `open()` and get that expected behavior.

**Comment by neothemachine** [comment]
I just find it a bit odd to point to something that doesn't exist yet. It's like saying to the browser cache, please give me "/firefox/cache/http_google.com" and then it would not say "sorry, doesn't exist" but actually download the document from http://google.com, save it, and then return it. So, in line with the principle separation of concerns, I think that the semantic identifier of a given ephemeris should be separate from how and where the data is stored. If one day you decide for example to actually store ephemeris data in a local database then this can be easily achieved by splitting those two concerns, but would be weird when you refer to a file path.

**Comment by tritium21** [comment]
On 11/26/2015 04:27, Maik Riechert wrote:
>
> I just find it a bit odd to point to something that doesn't exist yet. 
> It's like saying to the browser cache, please give me 
> "/firefox/cache/http_google.com" and then it would not say "sorry, 
> doesn't exist" but actually download the document from 
> http://google.com, save it, and then return it. So, in line with the 
> principle separation of concerns, I think that the semantic identifier 
> of a given ephemeris should be separate from how and where the data is 
> stored. If one day you decide for example to actually store ephemeris 
> data in a local database then this can be easily achieved by splitting 
> those two concerns, but would be weird when you refer to a file path.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/skyfielders/python-skyfield/issues/73#issuecomment-159856802>.
>
How do you write a new file?  You `open()` a file that doesn't exist.

**Comment by oefe** [comment]
Mixing source (the ephemeris name) and destination (directory/file) in the same parameter feels somewhat weird to me.

In the case of `open` you have only a single file, whether it is input or output, so there is no mixup.

**Comment by brandon-rhodes** [comment]
I like @neothemachine’s point about alternative non-filesystem storage systems, so I will go with the first option! Thanks for the input, everyone.

**Comment by brandon-rhodes** [comment]
All right! I have tried committing a `load()` routine with a `directory=` keyword, and I also went through and revamped some of the I/O logic. Anyone who is interested can uninstall Skyfield

    pip uninstall skyfield

and then try the development version

    pip install https://github.com/skyfielders/python-skyfield/archive/master.zip

if they want to go ahead and try it out. Hopefully I will push the next version late today or tomorrow in any case.

Thanks for the feature suggestion!
